Add ABI check for shims

unseddd opened the pull request:
Resolves #1631

RalfJung commented:
Thanks for the PR, @unseddd! However, we must have miscommunicated somewhere. This is not the check that I meant. In fact I think this check is redundant, since the same check is already done by the Miri core engine.

I probably should give an example for what I mean. Consider this code:
```rust
fn main() {
    extern "Rust" {
        fn malloc(size: usize) -> *mut std::ffi::c_void;
    }

    unsafe {
        let _ = malloc(0);
    };
}
```
This code is wrong, because it calls `malloc` with the "Rust" call ABI, when it should be "C". However, Miri accepts this code! https://github.com/rust-lang/miri/issues/1631 is about adjusting Miri to reject this code. The problem does not just affect `malloc`, it affects every single extern function that Miri has a "shim" for, i.e., that is supported by Miri.

I hope it's clearer now, and sorry for not giving a concrete example earlier. :) I will add this to the issue description as well.

unseddd commented:
> it calls malloc with the "Rust" call ABI, when it should be "C". However, Miri accepts this code! #1631 is about adjusting Miri to reject this code.

Ok, think I understand now. Honestly wasn't sure if the check(s) I PRed were correct in this context, so thanks for clarifying.

Would it be correct to just do something simple like:

```rust
if abi != Abi::C {
    return err...;
}
```

unseddd commented:
@RalfJung I changed the ABI check, and rebased on latest `master`. Let me know what you think.

With the update, looks like a test(s) is now [failing](https://github.com/rust-lang/miri/pull/1670/checks?check_run_id=1738314908#step:8:207). Is this expected? Should I fix the test, or did I screw something up?

RalfJung commented:
Sorry for the delay; I am quite busy currently so it could take a bit until I can take a closer look at what you did. It#s on my queue though and will not get lost!

> Would it be correct to just do something simple like

Well, the question is where you meant to put that check. The ABI might not be the same for all the shims we have. It probably is  `Abi::C` for most of them, but some functions will have other ABIs. Probably what makes most sense is to first handle the other ABIs, and then use the existing code for the C ABI.

RalfJung commented:
> With the update, looks like a test(s) is now failing. Is this expected? Should I fix the test, or did I screw something up?

That's a bit hard to tell, which function is it talking about? If it is `__rust_begin_short_backtrace`, then that function does have the Rust ABI, not the C ABI -- so looks like the check is wrong here.

unseddd commented:
> Sorry for the delay; I am quite busy currently so it could take a bit until I can take a closer look at what you did. It's on my queue though and will not get lost!

No worries, thanks for letting me know. No intention of being a nuisance :smile:

> If it is __rust_begin_short_backtrace, then that function does have the Rust ABI, not the C ABI -- so looks like the check is wrong here

I'll look into it more, then. Still need to figure out all the callers of the shim `find_mir_or_eval_fn`. From your earlier comment, I thought these would all be `extern "C"` calls, but maybe that's only a subset? Anyway, thanks for the feedback, will keep working on this.

RalfJung commented:
> Still need to figure out all the callers of the shim find_mir_or_eval_fn. 

`find_mir_or_eval_fn` is called *for each function call*. Usually it looks up the MIR of the function, but for some function, there's no MIR -- in particular, for `extern`-imported functions. Generally Miri will just error, saying that it cannot call functions for which there is no MIR available. However, for some specific of these functions, we have implemented "shims" by hand that implement the intended behavior by directly manipulating interpreter state.

The main entry point you have to worry about is this:

https://github.com/rust-lang/miri/blob/903bfd82bfe4fca0b85c8b7eb01b5841eac45f8c/src/shims/foreign_items.rs#L112

unseddd commented:
> find_mir_or_eval_fn is called for each function call. ... The main entry point you have to worry about is this

Thanks for pointing me in the right direction, that solved the failing test :smile:

Sorry for the notification spam in the related issue (forgot about the commit message). This PR should be ready for review when you are.

RalfJung commented on src/shims/foreign_items.rs:
```diff
@@ -165,6 +166,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             // Normally, this will be either `libpanic_unwind` or `libpanic_abort`, but it could
```
There are some more functions a bit further up this code that should also have their ABI checked.

RalfJung commented:
This function here should also check the ABI similar to what you did for `foreign_items`:

https://github.com/rust-lang/miri/blob/6fdda8aac97c57c84671ae54b6c9377245bc33e8/src/shims/dlsym.rs#L29

unseddd commented on src/shims/foreign_items.rs:
```diff
@@ -481,14 +508,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
             // Architecture-specific shims
             "llvm.x86.sse2.pause" if this.tcx.sess.target.arch == "x86" || this.tcx.sess.target.arch == "x86_64" => {
+                check_abi(abi, Abi::C)?;
```
Not sure if this is correct, or should be `Abi::X86Interrupt` (or smth?).

RalfJung replied:
Well, the test suite will tell. ;)

RalfJung commented on src/shims/foreign_items.rs:
```diff
@@ -504,6 +508,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             "llvm.x86.sse2.pause" if this.tcx.sess.target.arch == "x86" || this.tcx.sess.target.arch == "x86_64" => {
+                check_abi(abi, Abi::C)?;
                 let &[] = check_arg_count(args)?;
                 this.yield_active_thread();
```
I just added another architecture-specific function here; can you rebase and also add the `check_abi` here?

RalfJung commented:
Looking great. :) Please squash the commits together, then we can merge this.

(Heads-up, there might be conflicts with me resolving https://github.com/rust-lang/rust/issues/81341... sorry for that.)

bors commented:
:umbrella: The latest upstream changes (presumably #1688) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

unseddd commented:
> Looking great. :) Please squash the commits together, then we can merge this.

All done :smile:

> (Heads-up, there might be conflicts with me resolving rust-lang/rust#81341... sorry for that.)

No worries, was an easy fix.

RalfJung commented on src/shims/posix/foreign_items.rs:
```diff
@@ -14,12 +15,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     ) -> InterpResult<'tcx, bool> {
         let this = self.eval_context_mut();
 
+        check_abi(abi, Abi::C)?;
```
So, for foreign items you just enforce that all Posix symbols use the `C` ABI... can you do the same for `Dlsym`, instead of pushing things down? This reduces the risk of forgetting to add the check there. ;)

unseddd replied:
> So, for foreign items you just enforce that all Posix symbols use the C ABI... can you do the same for Dlsym, instead of pushing things down? This reduces the risk of forgetting to add the check there. ;)

Definitely ;) Wasn't sure if there would be non-C calls in either of the Posix Dlsyms. All fixed

RalfJung commented:
@unseddd thanks a lot for the PR and for patiently taking care of all my comments. :)
@bors r+

bors commented:
:pushpin: Commit de4eea9d49b5d4d5e10e67e1ae2754bbb720564c has been approved by `RalfJung`

<!-- @bors r=RalfJung de4eea9d49b5d4d5e10e67e1ae2754bbb720564c -->
<!-- homu: {"type":"Approved","sha":"de4eea9d49b5d4d5e10e67e1ae2754bbb720564c","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit de4eea9d49b5d4d5e10e67e1ae2754bbb720564c with merge b38dc837f2c2ccca6899d08956c94d6b682091ac...
<!-- homu: {"type":"BuildStarted","head_sha":"de4eea9d49b5d4d5e10e67e1ae2754bbb720564c","merge_sha":"b38dc837f2c2ccca6899d08956c94d6b682091ac"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/runs/1794140296)
Approved by: RalfJung
Pushing b38dc837f2c2ccca6899d08956c94d6b682091ac to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/runs/1794140296"},"merge_sha":"b38dc837f2c2ccca6899d08956c94d6b682091ac"} -->

unseddd commented:
> @unseddd thanks a lot for the PR and for patiently taking care of all my comments. :)

No problem :smile:

Thanks for being patient with me during the learning process :heart:
